ruby/examples/lib/apply_order_discount.rb

```diff
@@ -0,0 +1,44 @@
+class ApplyOrderDiscount
+  Request = Struct.new(:id)
+  Response = Struct.new(:discount, :has_discount?)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
```

Comment:
It will only have references to the concrete thing when you want it to, if you pass in an alternative implementation the class will never be initialised. Can't see any practical reasons not to go with this.